### ISSUE TITLE
Toggle code completion with a mappable key

### ISSUE THREAD

**Issue by maxlittlemore** [godotengine/godot-proposals#6656]
### Describe the project you are working on

Everything I work on. All. It's a request for the gdscript editor.

### Describe the problem or limitation you are having in your project

It's every project, working on gdscrip in particular.

My current example (not the only one, just the one that makes me swear the most):

When prototyping procedural mesh algorithms, particularly when building indices, I find myself building arrays of magic numbers with hard coded values.

If every time you enter a value like '2' at the end of a line and press the "Enter" key, even after a comma (,), you end up with a weird constant related to mouse positions rather than the value just entered. It kind of comes across as broken. It's also very annoying. And slow as a developer.

I shouldn't have to hit 'ESC' every time I enter a value in a list. I currently use vim to do these edits because it's quicker than your editor, but that sucks because I have to switch my own cognitive context between editors, and that makes using Godot feel like a chore.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Create a key mapping to toggle code completion. it could be defaulted to 'Alt' for Windows and RIGHT-CTL for Linux. (Not sure on Mac, not familiar with that environment since about 2009).

Note this is not the same as the current option to press 'ESC'. I shouldn't have to press  'ESC' then 'Enter' to get a new line without the most annoying pop up in the history of pop-ups popping up.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

With toggled to 'on':
Annoying pop-up while I'm writing code, hit that key, annoying pop-up goes away.
With toggled to 'off':
Struggling to remember a member of a built in, I put the cursor on the word and hit that key, I get the code completion thingy pop up.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This will be used often. No doubt. It's a quality of life issue. I know a similar thing was requested for G3, I know everyone that expects standard control from their own text editor will want this.

### Is there a reason why this should be core and not an add-on in the asset library?

It's the gdscript editor. built into the IDE.

**Comment by zinnschlag** [comment]
Assigning functions to modifier keys sounds problematic. Too easy to activate them accidentally while using another shortcut.  Also, I don't think Godot's input system supports binding functions to modifier keys in the first place. So it would have to be a regular shortcut instead.

As a workaround you can set the delay time for the pop up to maximum and then bind manual invocation of the pop up to a shortcut. That is the solution I am currently using. Doesn't work perfectly (pop up still sometimes steals my focus), but it is good enough.

**Comment by maxlittlemore** [comment]
Thanks zinschalag for your reply. All I'm asking for is a boolean switch in environment settings, a yeas/no on a switch of a 'feature'. For many of us, that 'feature' is a hindrance. It would be nice if I could stay within one environment....

**Comment by maxlittlemore** [comment]
or I guess I just use vim, does anyone have a vim setup for gdscript?

**Comment by KoBeWi** [comment]
What version are you using? Your problem sounds like a bug that was fixed recently.

**Comment by maxlittlemore** [comment]
v 3/0, 3.1, 3.2, 3.4, 3.5, 4.0

not exactly a bug, a feature request

**Comment by maxlittlemore** [comment]
To be clear, have a switch key to turn/off auto correct on or off. ( on windblows probs Alt, on Linux right-lctrl, on mac whatever(who cares?))

**Comment by KoBeWi** [comment]
No, autocompletion triggering on numbers was a bug: https://github.com/godotengine/godot/pull/74466
It's fixed in 4.0.1

**Comment by maxlittlemore** [comment]
Yes but...
Make it a thing that can be turned off/on with a mapped key....
That's my request
Is that so hard to understand?

**Comment by maxlittlemore** [comment]
like I said, the "problem" exists in:
3.0,3.2, 3.5. 4.0
I'm requesting a new feature.

**Comment by L4Vo5** [comment]
I have a similar issue where code completion insists on highlighting a different entry even after having typed a valid identifier.
For example, if you're in a Control node's script and type `var my_theme = theme`, then press enter to move to the next line, the first code completion suggestion will actually be the `theme_changed` signal, so it'll replace `theme` with that.
![imagen](https://user-images.githubusercontent.com/37083802/230749298-8c91a75c-f523-4a79-b0fd-0040108fe4d6.png)
But I think the solution for that should be to improve the default behaviour, not add a code completion key

**Comment by maxlittlemore** [comment]
I suppose my issue is I don't like using the mouse when I'm writing coding, and I have a brain so full of shortcut keys for a bazillion programs, sometimes I just like a simple editor where what I type is what is filed.

To that extent, I have considered asking for an optional 'vi' keyboard mode in the script editor for cases where there's a lot of repetition, but I know that'll never happen...

@L4Vo5 I have had this issue too, and that's a huge part of my irritation, but my request, as it seems to me, should be low hanging fruit.

My bad, I haven't yet taken the time to delve into the source around this yet, but symptomatically it seems the error detection is tightly coupled with the auto-completion.

I'm just asking for a flag that says "run that widget? Y/N". I've never seen something quite that simple in an editor, it's always too much or not enough.I really would love to see it.

**Comment by Calinou** [comment]
In 4.2.dev1 and later, you can prevent autocompletion from automatically popping up entirely with the setting added in https://github.com/godotengine/godot/pull/68140.
